### PR TITLE
Feature 5

### PR DESCRIPTION
**Pull request by Yehor40in**

### CODE REVIEW

File: MusicApp/Auxiliaries.swift [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    }
+    static var trackChangedNotification: Notification.Name {
+        return Notification.Name("trackChanged")
+    }
+    static var trackResumedNotification: Notification.Name {
+        return Notification.Name("trackResumed")
+    }
+    static var trackPausedNotification: Notification.Name {
+        return Notification.Name("trackPaused")
+    }
```
Comment (omaidanyk) [inline MusicApp/Auxiliaries.swift]: nit: You could write extension directly to Notification.Name and use it. Like:
```
extension Notification.Name {
    static var trackChanged: Notification.Name {
        return Notification.Name("trackChanged")
    }
}
```
usage looks more clear and understandable: `Notification.Name.trackChanged` instead of `Constants.trackChangedNotification`

File: MusicApp/ControlsViewController/ControlsViewController.swift [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+//  MusicApp
+//
+//  Created by Yehor Sorokin on 10/10/19.
+//  Copyright © 2019 Yehor Sorokin. All rights reserved.
+//
+
+import UIKit
+import MediaPlayer
+
+class ControlsViewController: UIViewController {
```
Comment (omaidanyk) [inline MusicApp/ControlsViewController/ControlsViewController.swift]: since you are not going to make a child classes better mark this one as final

File: MusicApp/ControlsViewController/ControlsViewController.swift [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    // MARK: - Outlets
+    @IBOutlet weak var songProgress: UIProgressView!
+    @IBOutlet weak var playButton: UIButton!
+    @IBOutlet weak var forwardButton: UIButton!
+    @IBOutlet weak var backwardButton: UIButton!
+    @IBOutlet weak var songName: UILabel!
+    @IBOutlet weak var artist: UILabel!
```
Comment (omaidanyk) [inline MusicApp/ControlsViewController/ControlsViewController.swift]: should be private, isn't it?

File: MusicApp/ControlsViewController/ControlsViewController.swift [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            selector: #selector(handlePlay(_:)),
+            name: Constants.trackPausedNotification,
+            object: nil
+        )
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(handlePaused(_:)),
+            name: Constants.trackResumedNotification,
+            object: nil
+        )
```
Comment (omaidanyk) [inline MusicApp/ControlsViewController/ControlsViewController.swift]: nit: usually in `viewDidLoad` could be another setup things. And if we don't want to overload `viewDidLoad` function, it's better to move such things into separate method right away. Smth like `setupTrackNotifications()` and call it here.

File: MusicApp/MusicListController/MusicListCell.swift [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+//
+
+import UIKit
+import MediaPlayer
+
+class MusicListCell: UITableViewCell {
+    // MARK: - Outlets
+    @IBOutlet weak var songCover: UIImageView!
+    @IBOutlet weak var songName: UILabel!
+    @IBOutlet weak var artist: UILabel!
```
Comment (omaidanyk) [inline MusicApp/MusicListController/MusicListCell.swift]: should be private

File: MusicApp/MusicListController/MusicListCell.swift [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    // MARK: - Properties
+    var item: MPMediaItem? {
+        didSet {
+            let img = item?.artwork?.image(at: songName.bounds.size)
+            self.songCover.image = img ?? UIImage(named: Constants.musicIconPlaceholderName)
+            self.songName.text = item?.title
+            self.artist.text = item?.artist ?? "Unknown"
```
Comment (omaidanyk) [inline MusicApp/MusicListController/MusicListCell.swift]: nit: it will work without `self.`. Am I wrong?

File: MusicApp/ControlsViewController/ControlsViewController.swift [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        delegate?.backward()
+    }
+    @IBAction func playTapped(_ sender: Any) {
+        delegate?.playPause()
+    }
+    @IBAction func forwardTapped(_ sender: Any) {
+        delegate?.forward()
+    }
+    @objc func handlePaused(_ notification: Notification) {
+        playButton.setImage(UIImage(named: "pause"), for: .normal)
```
Comment (omaidanyk) [inline MusicApp/ControlsViewController/ControlsViewController.swift]: better avoid direct usage of Strings. Such as you can use this image in another places in this controller it's better to create some computed property:
```
private var pauseButton: UIImage {
    return UIImage(named: "pause")
}
```

With computed property you can easily replace pause image with another one by just replacing only one string instead of looking for all occurrences along the file

Another solution can be: create private Constants enum in this ViewController with names (for images, notifications and so on). But you should be careful with static Stings because static variables increase the binary size

Also you can combine these two solutions and use private enum with static strings and calculated variables for images.

File: MusicApp/ControlsViewController/ControlsViewController.swift [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        updater?.isPaused = false
+    }
+    @objc func handlePlay(_ notification: Notification) {
+        playButton.setImage(UIImage(named: "play"), for: .normal)
+        updater?.isPaused = true
+    }
+    @objc func updateDetails(_ notification: Notification) {
+        updater?.invalidate()
+        if let item = notification.userInfo?["playingItem"] as? MPMediaItem {
```
Comment (omaidanyk) [inline MusicApp/ControlsViewController/ControlsViewController.swift]: using strings directly as a key for notifications is also bad practice.

File: MusicApp/ControlsViewController/ControlsViewController.swift [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+            case .playing:
+                playButton.setImage(UIImage(named: "pause"), for: .normal)
+                updater?.isPaused = false
+            default:
+                playButton.setImage(UIImage(named: "play"), for: .normal)
+            }
+        }
+    }
+    @objc func trackAudio() {
+        songProgress.progress += vps!
```
Comment (omaidanyk) [inline MusicApp/ControlsViewController/ControlsViewController.swift]: force unwrap is dangerous, better use `guard`. Why did swiftlint even allow it?

File: MusicApp/MusicListController/MusicListCell.swift [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+//  MusicApp
+//
+//  Created by Yehor Sorokin on 10/1/19.
+//  Copyright © 2019 Yehor Sorokin. All rights reserved.
+//
+
+import UIKit
+import MediaPlayer
+
+class MusicListCell: UITableViewCell {
```
Comment (omaidanyk) [inline MusicApp/MusicListController/MusicListCell.swift]: final

File: MusicApp/MusicListController/MusicListController.swift [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+//
+//  Created by Yehor Sorokin on 10/2/19.
+//  Copyright © 2019 Yehor Sorokin. All rights reserved.
+//
+
+import UIKit
+import MediaPlayer
+import StoreKit
+
+class MusicListController: UIViewController {
```
Comment (omaidanyk) [inline MusicApp/MusicListController/MusicListController.swift]: final

File: MusicApp/MusicListController/MusicListController.swift [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+    var sectionTitles: [String]?
+    var player: MPMusicPlayerController?
+    var query: MPMediaQuery?
+    // MARK: - Outlets
+    @IBOutlet weak var tableView: UITableView!
+    @IBOutlet weak var playingView: UIView!
+    @IBOutlet weak var playingCover: UIImageView!
+    @IBOutlet weak var playingName: UILabel!
+    @IBOutlet weak var playButton: UIButton!
+    @IBOutlet weak var forwardButton: UIButton!
```
Comment (omaidanyk) [inline MusicApp/MusicListController/MusicListController.swift]: all these things should be private as well.

File: MusicApp/MusicListController/MusicListController.swift [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+    // MARK: - Methods
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        SKCloudServiceController.requestAuthorization { status in
+            if status == .authorized {
+                self.query = MPMediaQuery.songs()
```
Comment (omaidanyk) [inline MusicApp/MusicListController/MusicListController.swift]: should `self` be weak in this closure? Better make `self` weak and move this preparation into separate method

File: MusicApp/MusicListController/MusicListController.swift [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+                self.items = self.preparedItems(from: (self.query?.items)!, by: .title)
+                self.preparePlayer()
+                DispatchQueue.main.async {
+                    self.tableView.dataSource = self
+                    self.tableView.delegate = self
```
Comment (omaidanyk) [inline MusicApp/MusicListController/MusicListController.swift]: why do you set it here? why can't you set dataSource and delegate directly in storyboard?

File: MusicApp/MusicListController/MusicListController.swift [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+
+        self.present(actionSheet, animated: true, completion: nil)
+    }
+
+}
+
+extension MusicListController: UITableViewDataSource {
+    // MARK: - Table view data source
+    func numberOfSections(in tableView: UITableView) -> Int {
+        return (sectionTitles?.count)!
```
Comment (omaidanyk) [inline MusicApp/MusicListController/MusicListController.swift]: force unwrap is dangerous. Better use `guard` or default value: `return sectionTitles?.count ?? 0`

File: MusicApp/MusicListController/MusicListController.swift [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+        navigationController?.navigationBar.prefersLargeTitles = true
+        playingView.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(self.showDetails(_:))))
+    }
+    // MARK: - Actions
+    @IBAction func playButtonTapped(_ sender: Any) {
+        if (player?.isPreparedToPlay)! && !forwardButton.isEnabled {
+            playRandomSong()
+        } else if player?.playbackState == .paused {
+            player?.play()
+            playButton.setImage(UIImage(named: "pause"), for: .normal)
```
Comment (omaidanyk) [inline MusicApp/MusicListController/MusicListController.swift]: that's what I was talking about images. Probably you should make this images as a computed properties in common `enum Constants`

File: MusicApp/MusicListController/MusicListController.swift [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+        } else {
+            player?.pause()
+            playButton.setImage(UIImage(named: "play"), for: .normal)
+        }
+    }
+    @IBAction func forwardTapped(_ sender: Any) {
+        playRandomSong()
+    }
+    @IBAction func sortTapped(_ sender: Any) {
+        let actionSheet = UIAlertController(title: nil, message: "Sort by", preferredStyle: .actionSheet)
```
Comment (omaidanyk) [inline MusicApp/MusicListController/MusicListController.swift]: using strings directly could be a problem when you will add multiple localisations. Better keep needed strings in one place along the file like private or public enum with names/titles/messages and so on.

File: MusicApp/MusicListController/MusicListController.swift [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+        actionSheet.view.tintColor = UIColor.green
+        actionSheet.addAction(UIAlertAction(title: "Artist", style: .default, handler: { (_) in
+            self.items = self.preparedItems(from: (self.query?.items)!, by: .artist)
+            DispatchQueue.main.async {
+                self.tableView.reloadData()
```
Comment (omaidanyk) [inline MusicApp/MusicListController/MusicListController.swift]: Does it complain that `self` should be weak?

File: MusicApp/MusicListController/MusicListController.swift [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+extension MusicListController: UITableViewDataSource {
+    // MARK: - Table view data source
+    func numberOfSections(in tableView: UITableView) -> Int {
+        return (sectionTitles?.count)!
+    }
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        let key = Character(sectionTitles![section])
+        return (items?[key]?.count)!
```
Comment (omaidanyk) [inline MusicApp/MusicListController/MusicListController.swift]: force unwrap!

File: MusicApp/MusicListController/MusicListController.swift [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+    }
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let key = Character(sectionTitles![indexPath.section])
+        let cell = tableView.dequeueReusableCell(withIdentifier: "MusicListCell", for: indexPath) as? MusicListCell
```
Comment (omaidanyk) [inline MusicApp/MusicListController/MusicListController.swift]: there is only one place where you can use force unwrap: for casting reusable cells with `as!`. But personally I prefer to use `guard` and return just empty cell in `else` closure

File: MusicApp/MusicListController/MusicListControllerDelegates.swift [hunk 21]
```diff
@@ -0,0 +1,26 @@
+//
+//  MusicListControllerDelegate.swift
```
Comment (omaidanyk) [inline MusicApp/MusicListController/MusicListControllerDelegates.swift]: nit: why did you move it to separate delegate if you could keep it in main file with `MusicListController`? Is main file already big enough?

File: MusicApp/MusicListController/MusicListControllerUtilities.swift [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+import UIKit
+
+extension MusicListController {
+    // MARK: - Utilities
+    func preparedItems(from raw: [MPMediaItem], by option: SortOption) -> [Character: [MPMediaItem]] {
+        var prepared = [Character: [MPMediaItem]]()
+        sectionTitles = [String]()
+        switch option {
+        case .artist:
+            let temp = raw.sorted { $0.artist! < $1.artist! }
```
Comment (omaidanyk) [inline MusicApp/MusicListController/MusicListControllerUtilities.swift]: does not make sense. You can create `prepared` dictionary with unsorted `raw` and then just sort a `sectionTitles`

We should speak about this function

File: MusicApp/MusicListController/MusicListControllerUtilities.swift [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+        player?.setQueue(with: self.query!)
+        player?.prepareToPlay()
+    }
+    func setPlayingItem(for path: IndexPath) {
+        let key = Character(sectionTitles![path.section])
+        player?.nowPlayingItem = self.items?[key]?[path.row]
+    }
+    func updatePlayingView() {
+        if let object = self.player?.nowPlayingItem {
+            let img = object.artwork?.image(at: self.playingCover!.bounds.size)
```
Comment (omaidanyk) [inline MusicApp/MusicListController/MusicListControllerUtilities.swift]: force unwrap

File: MusicApp/MusicListController/MusicListControllerUtilities.swift [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+            playingCover.image = img ?? UIImage(named: Constants.musicIconPlaceholderName)
+            playingName.text = object.title!
```
Comment (omaidanyk) [inline MusicApp/MusicListController/MusicListControllerUtilities.swift]: force unwrap. Do we even need unwrap here?

File: MusicApp/MusicListController/MusicListControllerUtilities.swift [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+                playButton.setImage(UIImage(named: "play"), for: .normal)
+            }
+            self.forwardButton.isEnabled = true
+        }
+    }
+    func playRandomSong() {
+
+        let sec = Int.random(in: 0..<self.sectionTitles!.count)
+        let key = sectionTitles![sec].first!
+        let row = Int.random(in: 0..<self.items![key]!.count)
```
Comment (omaidanyk) [inline MusicApp/MusicListController/MusicListControllerUtilities.swift]: force unwrap

File: MusicApp/MusicListController/MusicListControllerUtilities.swift [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+        updatePlayingView()
+    }
+    @objc func showDetails(_ sender: UITapGestureRecognizer!) {
+        if let details = storyboard?.instantiateViewController(withIdentifier: "detailInfo") as? PlayingViewController {
+            let pos = Position(
+                coverOut: tableView.frame.height + (navigationController?.navigationBar.frame.height)!,
+                imageOutBottom: playingCover.frame.height + playingCover.frame.origin.y,
+                imageOutTrailing: playingCover.frame.width + playingCover.frame.origin.x
+            )
+            details.prepared = PreparedData(image: self.view.makeScreenshot()!, player: player, outPosition: pos)
```
Comment (omaidanyk) [inline MusicApp/MusicListController/MusicListControllerUtilities.swift]: force unwrap

File: MusicApp/PlayingViewController/PlayingViewController.swift [hunk 27]
```diff
@@ -0,0 +1,116 @@
...
+//  MusicApp
+//
+//  Created by Yehor Sorokin on 10/10/19.
+//  Copyright © 2019 Yehor Sorokin. All rights reserved.
+//
+
+import UIKit
+import MediaPlayer
+
+class PlayingViewController: UIViewController {
```
Comment (omaidanyk) [inline MusicApp/PlayingViewController/PlayingViewController.swift]: final

File: MusicApp/PlayingViewController/PlayingViewController.swift [hunk 28]
```diff
@@ -0,0 +1,116 @@
...
+    @IBOutlet weak var chevron: UIButton!
+    @IBOutlet weak var fakeBackground: UIImageView!
+    // MARK: - Constraints
+    @IBOutlet weak var coverImageBottomConstraint: NSLayoutConstraint!
+    @IBOutlet weak var coverImageTrailingConstraint: NSLayoutConstraint!
+    @IBOutlet weak var coverViewTopConstraint: NSLayoutConstraint!
+    @IBOutlet weak var backgroundTopConstraint: NSLayoutConstraint!
+    @IBOutlet weak var backgroundLeadingConstraint: NSLayoutConstraint!
+    @IBOutlet weak var backgroundBottomConstraint: NSLayoutConstraint!
+    @IBOutlet weak var backgroundTrailingConstraint: NSLayoutConstraint!
```
Comment (omaidanyk) [inline MusicApp/PlayingViewController/PlayingViewController.swift]: should be private

File: MusicApp/PlayingViewController/PlayingViewController.swift [hunk 29]
```diff
@@ -0,0 +1,116 @@
...
+    // MARK: - Properties
+    var prepared: PreparedData?
+    var player: MPMusicPlayerController?
```
Comment (omaidanyk) [inline MusicApp/PlayingViewController/PlayingViewController.swift]: private?

File: MusicApp/PlayingViewController/PlayingViewController.swift [hunk 30]
```diff
@@ -0,0 +1,116 @@
...
+    override var preferredStatusBarStyle: UIStatusBarStyle {
+           return .lightContent
+    }
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        if let temp = prepared {
+            fakeBackground.image = temp.image
+            player = temp.player
+            let img = player?.nowPlayingItem?.artwork?.image(at: coverImageView!.bounds.size)
```
Comment (omaidanyk) [inline MusicApp/PlayingViewController/PlayingViewController.swift]: force unwrap

File: MusicApp/PlayingViewController/PlayingViewController.swift [hunk 31]
```diff
@@ -0,0 +1,116 @@
...
+        }
+        chevron.isHidden = true
+        update(with: player?.nowPlayingItem)
+    }
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        view.layoutIfNeeded()
+        coverViewTopConstraint.constant = (prepared?.outPosition.coverOut)!
+        coverImageBottomConstraint.constant = coverView.frame.height - (prepared?.outPosition.imageOutBottom)!
+        coverImageTrailingConstraint.constant = coverView.frame.width - (prepared?.outPosition.imageOutTrailing)!
```
Comment (omaidanyk) [inline MusicApp/PlayingViewController/PlayingViewController.swift]: force unwrap

File: MusicApp/PlayingViewController/PlayingViewController.swift [hunk 32]
```diff
@@ -0,0 +1,116 @@
...
+                "playingItem": item as Any,
+                "state": player?.playbackState as Any,
+                "progress": player?.currentPlaybackTime as Any
+            ]
+        )
+    }
+    func animateCoverIn() {
+        coverViewTopConstraint.constant = 50
+        coverImageBottomConstraint.constant = 20
+        coverImageTrailingConstraint.constant = 20
```
Comment (omaidanyk) [inline MusicApp/PlayingViewController/PlayingViewController.swift]: this default values looks like a "magic numbers". Better create some `let` constants for them. Or move them to some private enum like CoverConstraintsConstants with values fo "in" and "out" states

File: MusicApp/PlayingViewController/PlayingViewController.swift [hunk 33]
```diff
@@ -0,0 +1,116 @@
...
+        let verticalOffset = view.frame.height * 0.05
+        let horizontalOffset = view.frame.width * 0.05
+        backgroundTopConstraint.constant = verticalOffset
+        backgroundLeadingConstraint.constant = horizontalOffset
+        backgroundBottomConstraint.constant = verticalOffset
+        backgroundTrailingConstraint.constant = horizontalOffset
+        UIView.animate(withDuration: 0.3, animations: {
+            self.view.layoutIfNeeded()
```
Comment (omaidanyk) [inline MusicApp/PlayingViewController/PlayingViewController.swift]: weak self?

File: MusicApp/PlayingViewController/PlayingViewController.swift [hunk 34]
```diff
@@ -0,0 +1,116 @@
...
+        })
+    }
+    func animateCoverOut() {
+        coverViewTopConstraint.constant = (prepared?.outPosition.coverOut)!
+        coverImageBottomConstraint.constant = (coverView.frame.height - (prepared?.outPosition.imageOutBottom)!)
+        coverImageTrailingConstraint.constant = (coverView.frame.width  - (prepared?.outPosition.imageOutTrailing)!)
+        backgroundTopConstraint.constant = 0
+        backgroundLeadingConstraint.constant = 0
+        backgroundBottomConstraint.constant = 0
+        backgroundTrailingConstraint.constant = 0
```
Comment (omaidanyk) [inline MusicApp/PlayingViewController/PlayingViewController.swift]: move to private constants

File: MusicApp/PlayingViewController/PlayingViewController.swift [hunk 35]
```diff
@@ -0,0 +1,116 @@
...
+        UIView.animate(withDuration: 0.3, animations: {
+            self.view.layoutIfNeeded()
```
Comment (omaidanyk) [inline MusicApp/PlayingViewController/PlayingViewController.swift]: weak self?

File: MusicApp/PlayingViewController/PlayingViewControllerControlsDelegate.swift [hunk 36]
```diff
@@ -0,0 +1,33 @@
+//
+//  PlayingViewControllerControlsDelegate.swift
```
Comment (omaidanyk) [inline MusicApp/PlayingViewController/PlayingViewControllerControlsDelegate.swift]: own file is not necessary for this delegate but it's not a problem

File: MusicApp/MusicListController/MusicListController.swift (hunk 18 above)
Comment (Yehor40in) [reply MusicApp/MusicListController/MusicListController.swift]: No, compiler says it is ok

File: MusicApp/MusicListController/MusicListControllerUtilities.swift (hunk 22 above)
Comment (Yehor40in) [reply MusicApp/MusicListController/MusicListControllerUtilities.swift]: Yes, but sectionTitles is filled up depending on sorting option. So it seems to me that it is easier to get those values from temporary array (raw) rather than looping through dict and constructing a new sectionTitles, but I might be wrong about it

File: MusicApp/MusicListController/MusicListController.swift [hunk 37]
```diff
@@ -0,0 +1,231 @@
...
+                    }
+                    prepared[" "]!.append($0)
+                }
+            }
+        }
+        return prepared
+    }
+    func preparePlayer() {
+        player = MPMusicPlayerController.systemMusicPlayer
+        player?.setQueue(with: query!)
```
Comment (omaidanyk) [inline MusicApp/MusicListController/MusicListController.swift]: force unwrap
